### ISSUE TITLE
defineConstant setting should work with dotted attribute access

### ISSUE THREAD

**Issue by toofar** [microsoft/pyright#4060]
**Is your feature request related to a problem? Please describe.**

Currently we can use the `defineConstant` setting with conditionals on simple names like `DEBUG`, I would like to be able to use it with attributes of modules or objects too, like `config.DEBUG`.

Here is a reproducer using an object attribute. In our actual setup though we are using a module attribute, eg `import machinery; if machinery.USE_STANDARD: ...`:

```python
# Using with a config file like this I hope to guide the type checker to use a
# specific path of the conditional import below:
# {
#   "defineConstant": {
#     "USE_STANDARD": true,
#     "machinery.USE_STANDARD": true
#   }
# }

from dataclasses import dataclass
@dataclass
class Machinery:
    USE_STANDARD: bool

USE_STANDARD = True
machinery = Machinery(USE_STANDARD)

# if USE_STANDARD:  # <-- using a plain name in a conditional works
if machinery.USE_STANDARD:  # <-- using an attribute access does not, nodeType = ParseNodeType.MemberAccess
    from collections import deque
else:
    try:
        from doesntexist import deque
    except ImportError:
        pass

reveal_type(deque)
```

FWIW this works with mypy using `--always-true=USE_STANDARD`, I guess it just looks at the attribute name.

**Describe the solution you'd like**
In the example above I would like for pyright to not consider the import in the false branch of the conditional. Currently in that example the revealed type of `deque` is `Type[deque[_T@deque]] | Unknown | Unbound`. One or both of the Unknown or Unbound make pylance/vscode see the type as `Any` and not offer completion options.

**Additional context**

We have a setup where we support running an application with a couple of different sets of backing libraries. So we have conditional imports gated by configuration and also a couple of places throughout the code where we gate logic on the same configuration. Because we do that in a bunch of places its convenient to import that config from a common place. Additionally we support more than two configurations so there is a few different variables we might have to look at so it is convenient to just import the module and access attributes later rather than import all of the attributes from the module. For example [here](https://github.com/qutebrowser/qutebrowser/blob/8dd2bd5/tox.ini#L216) are all the constants we are passing to mypy. [Here](https://github.com/qutebrowser/qutebrowser/blob/4cf6aa8/qutebrowser/qt/webenginecore.py) is an example of where we are using the conditionals to gate imports.

**Comment by erictraut** [comment]
You've proposed that the `defineConstant` configuration support expressions rather than simple identifiers. That would violate core design principles in pyright and raise many questions like: "Why allow member access (dotted) expressions but not other expression forms?" and "Is it intended to be a textual match, a syntactic match, a semantic match?". So this particular solution isn't one that I would consider.

Let me try to understand your problem a bit deeper so we can explore other potential solutions.

In your example above, there is a global-level constant called `USE_STANDARD`. Why not use that for the conditional checks directly rather than using the constant to allocate an object and initialize an instance variable within that object and then using that instance variable in the conditional checks? In other words, why not simply replace `if machinery.USE_STANDARD` with `if USE_STANDARD`? If it's a constant, then there's on difference between the two. You can still define this constant in a central location and use it in multiple locations within your source base.

**Comment by erictraut** [comment]
I looked at mypy's implementation of `--always-true` and `--always-false`. It explicitly supports member access expressions that reference a member name in the `--always-true` or `--always-false` set. This behavior is applied only for conditional expressions and only along some code paths that evaluate conditional expression types, so it's pretty hacky and inconsistent.

It also strikes me as problematic to assume that the behavior should apply to all namespaces rather than limiting it to identifiers in a global namespace.

Pyright could replicate mypy's behavior here, but I have concerns. I'll need to think about this more.

I'm still interested in the answer to the question I posed above.

**Comment by toofar** [comment]
Both the object allocation and the handy constant in the same file where just to keep the example within a single file. Here's an example of what we are doing currently:
* a file that uses these conditionals: https://github.com/qutebrowser/qutebrowser/blob/4cf6aa8/qutebrowser/qt/webenginecore.py
* the machinery module that sets some common variables at runtime: https://github.com/qutebrowser/qutebrowser/blob/4cf6aa8cef5227d3610b31144318cd78109d2809/qutebrowser/qt/machinery.py

So in this case they are actually module level variables that are set at runtime, and we override them for the type checker and then do one run per supported configuration to make sure we get good coverage.

How we are then accessing the variables currently is:

```python
from qutebrowser.qt import machinery

if machinery.USE_PYSIDE6:
   ...
```

We could change the import to be `from qutebrowser.qt.machinery import USE_PYSIDE6, ...` and that seems to work fine but it is a convention in the codebase to import modules and then access the attributes on them so that you can see where variables are coming from where they are used.

**Comment by erictraut** [comment]
After thinking about this further, I still have some concerns about mypy's behavior. But given that this is a relatively advanced feature and will have no effect unless you use the `defineConstant` configuration, it's unlikely for the inconsistency and namespace issues to cause broad problems. Presumably, those who use this feature will be capable of diagnosing and mitigating any issue that they encounter.

I've therefore added support for member access expressions so it works like mypy. This will be included in the next release of pyright.

**Comment by toofar** [comment]
Thank you very much for humouring me!

**Comment by erictraut** [comment]
This is included in pyright 1.1.276, which I just published. It will also be included in a future release of pylance.
